File: python/프로그래머스/지게차와크레인.py
from collections import deque
def 창고외부체크_DFS(storage, visited, start, moves):
    n = len(storage)
    m = len(storage[0])
    check = [[False] * m for _ in range(n)]
    
    def dfs(x, y):
        for dx, dy in moves:
            nx, ny = x + dx, y + dy
            # 창고 범위를 벗어나면 외부로 나갔다고 판단하여 True 반환
            if not (0 <= nx < n and 0 <= ny < m):
                return True
            # 이동 가능한 칸(visited가 1)이고 아직 방문하지 않은 경우
            if visited[nx][ny] == 1 and not check[nx][ny]:
                check[nx][ny] = True
                if dfs(nx, ny):
                    return True
        return False
    
    return dfs(start[0], start[1])

def 지게차(storage,target,moves, visited):
    n = len(storage)
    m = len(storage[0])
    visit  = [[0 for _ in range(m)] for _ in range(n)]

    temp = []
    for x in range(n):
        for y in range(m):
            if storage[x][y] == target:
                for move in moves:
                    nx, ny = x + move[0], y + move[1]
                    if 0 <= nx < n and 0 <= ny < m:
                        if visited[nx][ny] == 1:
                            if 창고외부체크_DFS(storage, visited, [nx,ny], moves):       
                                temp.append([x,y])
                                break
                    else:
                        temp.append([x,y])
                        break  
    for t in temp:
        visited[t[0]][t[1]] = 1

def 크레인(storage,target,moves, visited):
    n = len(storage)
    m = len(storage[0])
    for x in range(n):
        for y in range(m):
            if storage[x][y] == target:
                visited[x][y] = 1
                
def solution(storage, requests):
    moves = [(1,0),(-1,0),(0,1),(0,-1)]
    n = len(storage)
    m = len(storage[0])
    answer = 0
    visited = [[0 for _ in range(m)] for _ in range(n)]
    for req in requests:
        req_li = list(req)
        if len(req_li) == 1:
            지게차(storage, req_li[0], moves, visited)
        else:
            크레인(storage, req_li[0], moves, visited)
    
    for a in visited:
        for b in a:
            if b == 0:
                answer += 1
    
    return answer

File: java/지게차와크레인.java
import java.util.*;

class Solution {
    private String[][] STORAGE = {{}};
    private boolean[][] visited = {{}};
    private int[][] next = {{1,0},{-1,0},{0,1},{0,-1}};
    private boolean isBlocked = false;
    
    public int solution(String[] storage, String[] requests) {
        int answer = 0;
        STORAGE = new String[storage.length][]; 
        visited = new boolean[storage.length][];
        
        for (int i = 0; i < storage.length; i++) {
            STORAGE[i] = storage[i].split(""); 
            visited[i] = new boolean[STORAGE[i].length];
        }
        
        for(String request: requests) {
            if(request.length() == 2) {
                크레인(request.split("")[0]);
                continue;
            } 
            
            List<int[]> visit = new ArrayList<>();
            for(int i=0;i<STORAGE.length;i++) {
                for(int j=0;j<STORAGE[0].length;j++) {
                    if(!visited[i][j] && STORAGE[i][j].equals(request)) {
                        boolean [][]temp = new boolean[visited.length][visited[1].length];
                        dfs(i,j,temp);
                        if(isBlocked) {
                            visit.add(new int[]{i,j});
                            isBlocked = false;
                        }
                    }
            
                }
            }
            
            for(int[]v: visit) {
                visited[v[0]][v[1]] = true;
            }
        }
        
        for(boolean[]vv: visited) {
            for(boolean v: vv) {
                System.out.print(v +" ");
                if (!v) {
                    answer += 1;
                }
            }
            System.out.println("");
        }
        return answer;
    }
    
    private void 크레인(String container) {
        for(int i=0;i<STORAGE.length;i++) {
            for(int j=0;j<STORAGE[i].length;j++) {
                if(!visited[i][j] && STORAGE[i][j].equals(container)) {
                    visited[i][j] = true;
                }
            }
        }
    }
    
    private void dfs(int x,int y, boolean[][] temp) {
        temp[x][y] = true;
        for (int[]nn: next) {
            int nx = x + nn[0];
            int ny = y + nn[1];
            
            if (0<=nx && nx<STORAGE.length && 0 <= ny && ny < STORAGE[0].length) {
                if(visited[nx][ny] && !temp[nx][ny]) {
                    dfs(nx,ny,temp);
                }
            } else {
                isBlocked = true;
                return;
            }
        }
    }
    
    private boolean bfs(int x, int y) {
        Deque<int[]> q = new ArrayDeque<>();
        q.addLast(new int[]{x,y});
        
        while(!q.isEmpty()) {
            int[] current = q.pollFirst();
            
            for(int[]nn: next) {
                int nx = current[0]+nn[0];
                int ny = current[1]+nn[1];
                            
                if (0<=nx && nx<STORAGE.length && 0 <= ny && ny < STORAGE[0].length) {
                    if(visited[nx][ny]) {
                        q.addLast(new int[]{nx,ny});   
                    }
                } else {
                    return true;
                }
            }
        }
        return false;
    }
}

